numpyro/distributions/discrete.py

```diff
@@ -301,6 +303,43 @@ def Categorical(probs=None, logits=None, validate_args=None):
         raise ValueError('One of `probs` or `logits` must be specified.')
 
 
+@copy_docs_from(Distribution)
+class Delta(Distribution):
+    arg_constraints = {'value': constraints.real, 'log_density': constraints.real}
+    support = constraints.real
+
+    def __init__(self, value=0., log_density=0., event_dim=0, validate_args=None):
+        if event_dim > np.ndim(value):
+            raise ValueError('Expected event_dim <= v.dim(), actual {} vs {}'
+                             .format(event_dim, np.ndim(value)))
+        batch_dim = np.ndim(value) - event_dim
+        batch_shape = np.shape(value)[:batch_dim]
+        event_shape = np.shape(value)[batch_dim:]
+        self.value = lax.convert_element_type(value, xla_bridge.canonicalize_dtype(np.float64))
+        # NB: following Pyro implementation, log_density should be broadcasted to batch_shape
+        self.log_density = promote_shapes(log_density, shape=batch_shape)[0]
+        super(Delta, self).__init__(batch_shape, event_shape, validate_args=validate_args)
+
+    def sample(self, key, sample_shape=()):
+        shape = sample_shape + self.batch_shape + self.event_shape
+        return np.broadcast_to(self.value, shape)
+
+    def log_prob(self, x):
+        if self._validate_args:
+            self._validate_sample(x)
+        log_prob = np.log(x == self.value)
+        log_prob = sum_rightmost(log_prob, len(self.event_shape))
+        return log_prob + self.log_density
```

Comment:
~Shouldn't `self.log_density` be only added to elements where `x == self.value`?~ I see that we are relying on inf addition here. Does that give any issues with gradients?

Reply:
We don't have tests for grad of invalid samples (for other distributions too). I am not sure how to properly test it.

Reply:
The following script shows that `grad` does not affect by `inf`, but I am not sure if that is what you want to test?
```
def f(x):
    return np.log(x == x + 1) + 2 * x

grad(f)(1.)  # 2
```

Reply:
My initial thought was that the gradient should be nan for values that don't match. i.e. suppose the initial value is `[1, 2]` (assuming event_ndim=0) and the value to be scored is `[1, 3]`, then the `log_prob` term would be `[0, -inf]`, and hence the gradient would be `[.., nan]`. However, this isn't correct either, and it probably doesn't matter because autograd will give the wrong value for the gradient.

Reply:
Got it. Hopefully we don't have usage cases which have invalid values for Delta distribution.